coverage/lcovreport.py

```diff
@@ -0,0 +1,106 @@
+# Licensed under the Apache License: http://www.apache.org/licenses/LICENSE-2.0
+# For details: https://github.com/nedbat/coveragepy/blob/master/NOTICE.txt
+
+"""LCOV reporting for coverage.py"""
+
+import sys
+import base64
+from hashlib import md5
+
+from coverage.report import get_analysis_to_report
+
+
+class LcovReporter:
+    """A reporter for writing LCOV coverage reports"""
+
+    report_type = "LCOV report"
+
+    def __init__(self, coverage):
+        self.coverage = coverage
+        self.config = self.coverage.config
+
+    def report(self, morfs, outfile=None):
+        """Renders the full lcov report
+
+        'morfs' is a list of modules or filenames
+
+        outfile is the file object to write the file into
+        """
+
+        self.coverage.get_data()
+        outfile = outfile or sys.stdout
+
+        for fr, analysis in get_analysis_to_report(self.coverage, morfs):
+            self.get_lcov(fr, analysis, outfile)
+
+    def get_lcov(self, fr, analysis, outfile=None):
+        """Produces the lcov data for a single file
+
+        get_lcov currently supports both line and branch coverage,
+        however function coverage is not supported.
+
+        """
+
+        outfile.write("TN:\n")
+        outfile.write(f"SF:{fr.relative_filename()}\n")
+        sourceLines = fr.source().splitlines()
+        for covered in sorted(list(analysis.executed)):
+            # Note: Coveragepy currently only supports checking *if* a line has
+            # been executed, not how many times, so we set this to 1 for nice
+            # output even if it's technically incorrect
+
+            # The lines below calculate a 64 bit encoded md5 hash of the line
+            # corresponding to the DA lines in the lcov file,
+            # for either case of the line being covered or missed in Coveragepy
+            # The final two characters of the encoding ("==") are removed from
+            # the hash to allow genhtml to run on the resulting lcov file
+            try:
+                line = sourceLines[covered - 1].encode("utf-8")
+            except IndexError:
+                line = b""
+            hashed = str(base64.b64encode(md5(line).digest())[:-2], encoding="utf-8")
```

Comment:
I guess tastes vary, but I like `.strip("=")` here more than `[:-2]` since it's more explicit what it's doing.
And "ascii" as encoding also seems more appropriate.

Reply:
But this is where we need to decide what to do about md5: it's not available on FIPS-enabled systems.  Here are some choices:
1. Don't use md5 at all, and leave the hash out of the report.
2. Use md5, but catch the ValueError it will raise on FIPS systems (https://github.com/nedbat/coveragepy/issues/1189)
3. Make lcov unavailable on FIPS systems
4. Let lcov just fail if used on FIPS systems

I don't know what the effect would be of omitting the hash, so I don't know which to pick.

Reply:
The hash is mainly used to ensure the correct lines are marked as covered, to my knowledge. Upstream's `genhtml`, for example, which uses an lcov file and the original source tree to produce an html page with annotated source files, will throw an error if it finds a hash that mismatches: https://github.com/linux-test-project/lcov/blob/master/bin/genhtml#L5308.

Genhtml is of course an unlikely usecase with pycoverage, but it's the reference consumer.

It is up to the implementation of the consumer to define what the md5 is used for, but it is marked as optional, so any tools that break on a missing md5 are inherently broken. There shouldn't be any issue with simply omitting the hash when it can't be set, and I don't think I can think of many other uses than "has our user accidentally changed their source". I think that was probably a more common use case around the time lcov added that feature :)

It *would* be sad if pycoverage produced different outputs depending on the machine it is run on. Maybe hence it's best to remove this functionality in general.

I think a fifth option is to lobby upstream for supporting a more modern hash, as much as that seems unnecessary. It's honestly a bit silly that python mandates thinking about FIPS compliance with no opt-out if you use its md5 hash, and even more silly that some default python distributions seem to enable FIPS compliance (this seems to be a MacOS thing?).

It'll be hard to come up with a backwards-compatible design for that, so I suspect the solution would be to deprecate having hashes at all, since those don't really seem that useful anymore in 2022 when checking for accidental source changes is anyway so easy and often unnecessary.

Reply:
Thanks for the backstory on the hash values.

> It's honestly a bit silly that python mandates thinking about FIPS compliance with no opt-out if you use its md5 hash

Is this Python's fault? FIPS forbids md5 as an algorithm.

> some default python distributions seem to enable FIPS compliance (this seems to be a MacOS thing?)

I haven't encountered this.  Can you provide more information?

Reply:
> Is this Python's fault? FIPS forbids md5 as an algorithm.

No, not at all. I just think the situation is a bit silly for backwards-compatibility with older formats that use md5 for legitimate reasons (or even just for migration purposes). It's besides the point, sorry for venting here.

> I haven't encountered this. Can you provide more information?

[This support page](https://support.apple.com/en-us/HT209638) seems to suggest a FIPS-compliance mode is enabled by default. Reading further, it may come down to processor support for the hash, rather than something explicitly disabled on the software side, but this is all ultimately besides the point.

I think I'll spend a bit more time reading up on where exactly the limitation comes from, and then raise an issue on the lcov upstream to see if they're interested enough in allowing lcov to be run in FIPS-compliant environments. No need for coveragepy to be involved in that :)

Reply:
@TLATER is right here on the use of the hash within LCOV. I am perfectly happy to remove the md5 hash if you would prefer @nedbat.